core/src/main/java/com/google/zxing/aztec/AztecWriter.java

```diff
@@ -39,7 +38,7 @@ public BitMatrix encode(String contents, BarcodeFormat format, int width, int he
 
   @Override
   public BitMatrix encode(String contents, BarcodeFormat format, int width, int height, Map<EncodeHintType,?> hints) {
-    Charset charset = StandardCharsets.ISO_8859_1;
+    Charset charset = null; // Do not add any ECI code by default
```

Comment:
One thing that might be worth highlighting here is that this PR makes the semantics of the Aztec `Encoder`/`AztecWriter` slightly different from QR/PDF417/DataMatrix in terms of character set handling: specifying `charset == null` or no encoding hint means "do not add any ECI code; compliant decoders should treat as ISO_8859_1 because that's what the spec says."

The QR decoder in particular differs from this…

https://github.com/zxing/zxing/blob/0cf3b9be71680f50c90a71ca26ce0d33664b0dd6/core/src/main/java/com/google/zxing/qrcode/decoder/DecodedBitStreamParser.java#L235-L244

_Changing_ the behavior of the QR decoder would probably break a _ton_ of software out there which relies on ZXing's behavior of guessing the encoding, so I'm certainly not suggesting it :sweat_smile:.

If anything, it _might_ be a good idea to add the _same_ encoding-guessing behavior for PDF417/DataMatrix/Aztec, in case there are broken encoders out there, which encode UTF-8 without ECI.

Reply:
Decoding is different from encoding, and that's a decoder you quote here. There, yes, I don't want to change the logic, but, it's also fine to take best guesses instead of failing.

Reply:
I don't know whether Aztec requires an ECI; if so and it's missing then that's reasonably an error. QR codes don't, so need to try something when not specified. Nevertheless, sure, sometimes it's useful to try to recover when an ECI really should be there but isn't, hence the attempt to guess common encodings.

Reply:
Aztec Code spec does not _require_ an ECI. Like most (all?) other 2D barcodes it's supposed to be read as ISO-8859-1 in the absence of an ECI.

> QR codes don't, so need to try something when not specified.

[As I understand it](https://www.linkedin.com/pulse/enhanced-channel-interpretation-terry-burton/), QR is also _supposed to be_ interpreted as ISO-8859-1 in the absence of ECI… but many encoders write UTF-8 without an ECI mark.

>  Nevertheless, sure, sometimes it's useful to try to recover when an ECI really should be there but isn't, hence the attempt to guess common encodings.

Yep, exactly.